### PR TITLE
plugin rake-fast: show task descriptions in autocomplete

### PR DESCRIPTION
**Pull request by allenan**
This PR adds rake task descriptions to the autocomplete.

**Before:**
screenshot 2015-12-21 13 17 26

**After:**
screenshot 2015-12-21 17 15 13

It saves both task names and descriptions to the .rake_tasks file, in a format that [_describe](http://zsh.sourceforge.net/Doc/Release/Completion-System.html#index-_005fdescribe) can understand.

Mind taking a look, @KevinBongart?

### CODE REVIEW

**Comment by KevinBongart** [comment]
What does it look like with tasks that don't have a description?

(I vaguely remember a PR for making rake-fast include tasks without description, but I can't remember if that ever got merged)

**Comment by allenan** [comment]
ah good point. Let me take a look at that case






On Tue, Dec 22, 2015 at 8:23 AM -0800, "Kevin Bongart" <notifications@github.com> wrote:












What does it look like with tasks that don't have a description?



(I vaguely remember a PR for making rake-fast include tasks without description, but I can't remember if that ever got merged)

**Comment by allenan** [comment]
It looks like it ignores tasks without descriptions since `rake --tasks` does not output tasks without descriptions. `rake --tasks --all` will show all tasks, which includes a lot of Rails tasks that don't have descriptions. 

Since this PR's behavior is identical to the existing state of rake-fast (just showing descriptions), I might recommend leaving this PR as is, and I can open another one that includes all tasks, if that's something we want.

File: plugins/rake-fast/rake-fast.plugin.zsh [hunk 1]
```diff
@@ -21,7 +26,9 @@ _rake () {
       echo "\nGenerating .rake_tasks..." > /dev/stderr
       _rake_generate
     fi
-    compadd `cat .rake_tasks`
+    local -a rake_options
+    rake_options=( "${(f)mapfile[.rake_tasks]}" )
```
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: Line

**Comment by tankhanleng** [comment]
Pull

File: plugins/rake-fast/rake-fast.plugin.zsh (hunk 1 above)
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: Pull

File: plugins/rake-fast/rake-fast.plugin.zsh [hunk 2]
```diff
@@ -12,7 +12,12 @@ _rake_does_task_list_need_generating () {
 }
 
 _rake_generate () {
-  rake --silent --tasks | cut -d " " -f 2 > .rake_tasks
+  rake --silent --tasks \
+    | sed "s/^rake //"  \
```
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: Pull

File: plugins/rake-fast/rake-fast.plugin.zsh [hunk 3]
```diff
@@ -21,7 +26,9 @@ _rake () {
       echo "\nGenerating .rake_tasks..." > /dev/stderr
       _rake_generate
     fi
-    compadd `cat .rake_tasks`
+    local -a rake_options
+    rake_options=( "${(f)mapfile[.rake_tasks]}" )
+    _describe 'values' rake_options
```
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: Pull

File: plugins/rake-fast/rake-fast.plugin.zsh (hunk 1 above)
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: Pull

File: plugins/rake-fast/rake-fast.plugin.zsh [hunk 4]
```diff
@@ -12,7 +12,12 @@ _rake_does_task_list_need_generating () {
 }
 
 _rake_generate () {
-  rake --silent --tasks | cut -d " " -f 2 > .rake_tasks
+  rake --silent --tasks \
+    | sed "s/^rake //"  \
+    | sed "s/\:/\\\:/g" \
+    | sed "s/\[.*\]//g" \
```
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: Sed

File: plugins/rake-fast/rake-fast.plugin.zsh (hunk 2 above)
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: sed

File: plugins/rake-fast/rake-fast.plugin.zsh [hunk 5]
```diff
@@ -12,7 +12,12 @@ _rake_does_task_list_need_generating () {
 }
 
 _rake_generate () {
-  rake --silent --tasks | cut -d " " -f 2 > .rake_tasks
+  rake --silent --tasks \
```
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: silent

Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: taske

Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: silent

Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: tasks

File: plugins/rake-fast/rake-fast.plugin.zsh [hunk 6]
```diff
@@ -12,7 +12,12 @@ _rake_does_task_list_need_generating () {
 }
```
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: tasks

Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: tasks

File: plugins/rake-fast/rake-fast.plugin.zsh [hunk 7]
```diff
@@ -12,7 +12,12 @@ _rake_does_task_list_need_generating () {
...
```
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: Tasks

File: plugins/rake-fast/rake-fast.plugin.zsh (hunk 6 above)
Comment (tankhanleng) [inline plugins/rake-fast/rake-fast.plugin.zsh]: rake_tasks

**Comment by KevinBongart** [comment]
@allenan Good idea. And last question: would this be a graceful transition? It looks like the `.rake_tasks` content gets a new format, so what happens when:

1. User has a `.rake_tasks` file, formatted with current format (no descriptions)
2. This PR gets merged
3. User updates OMZ
4. User types `rake` then tab

Ideally, it should refresh the `.rake_tasks`, but at the very least, it should still work (without the descriptions) until something prompts a `rake_refresh`.

**Comment by KevinBongart** [comment]
Thanks @tankhanleng for these thoughtful comments! :heart:

**Comment by allenan** [comment]
@KevinBongart good call. I just added a commit that checks the version of the format of .rake_tasks. This will cause it to regenerate if no version stamp is present or if the version stamp in the .rake_tasks file is behind the version in the code. let me know if you think this will fit the bill

**Comment by KevinBongart** [comment]
Seems like a nice strategy, though I had a strange output and result when trying locally (on a `.rake_tasks` file generated with the current master version of the plugin):

```zsh
$ rake ......
Generating .rake_tasks...h
_rake_generate:1: command not found: _rake_tasks_version

_rake:7: bad floating point constant

Generating .rake_tasks...
_rake_generate:1: command not found: _rake_tasks_version
_rake:7: bad floating point constant

Generating .rake_tasks...
_rake_generate:1: command not found: _rake_tasks_version
$ cat .rake_tasks
version:
about:...
acts_as_taggable_on_engine\:install\:migrations:...
acts_as_taggable_on_engine\:tag_names\:collate_bin:...
acts_as_taggable_on_engine\:tag_names\:collate_ci:...
affiliate\:track_conversions:...
[...]
```

Is this happening on your side as well? Interested in finding out what's wrong

**Comment by allenan** [comment]
That's really strange. I wasn't able to reproduce that exact error on my side. Although I did see that I missed the `shift rake_options` line. I rebased with that change. let me know if that helps anything on your end

**Comment by robbyrussell** [comment]
⌛️ Closing as this has been lingering for too many years. Happy to reconsider one day, though!

**Pull request closed without merging**